examples/udp_echo_server/main.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2008, 2009, 2010  Kaspar Schleiser <kaspar@schleiser.de>
+ * Copyright (C) 2013 INRIA
+ * Copyright (C) 2013 Ludwig Knüpfer <ludwig.knuepfer@fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     examples
+ * @{
+ *
+ * @file
+ * @brief       Default application that shows a lot of functionality of RIOT
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Oliver Hahm <oliver.hahm@inria.fr>
+ * @author      Ludwig Knüpfer <ludwig.knuepfer@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#define CONN_HAS_IPV6
+
+#include "net/af.h"
+#include "net/conn/udp.h"
+#include "msg.h"
+#include "shell.h"
+#include "shell_commands.h"
+#include "thread.h"
+
+#define ECHO_SERVER_BUFSIZE         (64)
+#define ECHO_SERVER_MSG             (0x9999)
+#define ECHO_SERVER_PORT            (61616)
+#define ECHO_SERVER_PRIO            (THREAD_PRIORITY_MAIN - 1)
+#define ECHO_SERVER_MSG_QUEUE_SIZE  (8)
+#define ECHO_SERVER_STACK_SIZE      (THREAD_STACKSIZE_MAIN)
+
+/* shell line buffer */
+static char line_buf[SHELL_DEFAULT_BUFSIZE];
+static char echo_server_buf[ECHO_SERVER_BUFSIZE];
+static char echo_server_stack[ECHO_SERVER_STACK_SIZE];
+static msg_t echo_server_msg_queue[ECHO_SERVER_MSG_QUEUE_SIZE];
+static const conn_ep_udp_t server_address = { .addr = { .ipv6 = IPV6_ADDR_UNSPECIFIED },
+                                              .family = AF_INET6,
+                                              .netif = CONN_EP_ANY_NETIF,
+                                              .port = ECHO_SERVER_PORT };
+static kernel_pid_t echo_server_pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief   Callback for server to react to received UDP packets.
+ *
+ * @param[in] conn  The connectivity object the packet was received on
+ */
+static void server_cb(conn_udp_t *conn)
+{
+    msg_t msg = { .type = ECHO_SERVER_MSG, .content = { .ptr = conn } };
+    /* inform server thread about reception */
+    msg_send(&msg, echo_server_pid);
+}
+
+static void *echo_server(void *arg)
+{
+    conn_udp_t server;
+
+    (void)arg;
+    /* initialize message queue for asynchronous IPC */
+    msg_init_queue(echo_server_msg_queue, ECHO_SERVER_MSG_QUEUE_SIZE);
+    /* create end-point for server */
+    if (conn_udp_create(&server, &server_address, server_cb) < 0) {
+        puts("Unable to start server");
+        return NULL;
+    }
+    while (1) {
+        msg_t msg;
+        conn_ep_udp_t client_address;
+        int res;
+
+        msg_receive(&msg);  /* wait for message */
```

Comment:
For none-asynchronous handling of incoming messages one could just start from the code from 88 here and remove the surrounding switch-case.